intermediate_source/parametrizations.py

```diff
@@ -0,0 +1,353 @@
+controls the Lipschitz constant of the network by dividing its parameters by
+their `spectral norm <https://en.wikipedia.org/wiki/Matrix_norm#Special_cases>`_,
+rather than its Frobenius norm.
+
+All these methods have a pattern in common. They all transform a parameter
+in an appropriate way before using it. In the first case, they make it orthogonal by
+using a function that maps matrices to orthogonal matrices. In the case of weight
+and spectral normalization, they divide the original parameter by its norm.
+
+More generally, all these examples use a function to put extra structure on the parameters.
+In other words, they use a function to constrain the parameters.
+
+In this tutorial, you will learn how to implement and use this pattern to write and
+put constraints on your model. Doing so is as easy as writing your own ``nn.Module``.
+
+Requirements
+------------
+``"torch>=1.9.0a0+7aeee28"``
+
+Implementing Parametrizations by Hand
+-------------------------------------
+
+Assume that we want to have a square linear layer with symmetric weights, that is,
+with weights :math:`X` such that :math:`X = X^{\intercal}`. One way to do so is
+to copy the upper triangular part of the matrix into its lower triangular part
+"""
+
+import torch
+import torch.nn as nn
+import torch.nn.utils.parametrize as P
+
+def symmetric(X):
+    return X.triu() + X.triu(1).transpose(-1, -2)
+
+X = torch.rand(3, 3)
+A = symmetric(X)
+print(A)
+assert torch.allclose(A, A.T)
+
+###############################################################################
+# We can then use this idea to implement a linear layer with symmetric weights:
+class LinearSymmetric(nn.Module):
+    def __init__(self, n_features):
+        super().__init__()
+        self.weight = nn.Parameter(torch.rand(n_features, n_features))
+
+    def forward(self, x):
+        A = symmetric(self.weight)
+        return x @ A
+
+###############################################################################
+# The layer can be then used as a regular linear layer
+layer = LinearSymmetric(3)
+out = layer(torch.rand(8, 3))
+
+###############################################################################
+# This implementation, although correct and self-contained, presents a number of problems:
+#
+# 1) It reimplements the layer. We had to implement the linear layer as ``x @ A``. This is
+#    not very problematic for a linear layer, but imagine having to reimplement a CNN or a
+#    Transformer...
+# 2) It does not separate the layer and the parametrization.  If the parametrization were
+#    more difficult, we would have to rewrite its code for each layer that we want to use it
+#    in.
+# 3) It recomputes the parametrization everytime forward is called. If we used the layer
+#    several times during the forward pass, (imagine the recurrent kernel of an RNN) we would
+#    be recomputing the same ``A`` every time the layer is called.
+#
+# Parametrizations come to solve all these and other problems.
```

Comment:
This sentence is actually attached to the above paragraph in the rendered version. Which doesn't look great.
Maybe we should move it as the first sentence of the next section?